### PR TITLE
Adds an uptime check

### PR DESCRIPTION
**Pull request by asachs01**
`check_uptime` provides the capability to have a native uptime check that is compatible across platforms. I'm currently still working on the check, but should have this nailed down later today.

### CODE REVIEW

**Comment by jkerry** [comment]
I'll take a look this morning. A lot of new code got merged last night ( though the master build passed )

**Comment by asachs01** [comment]
Looks like the lack of the check building was a failure on my part to add the check to the godel yaml.

**Comment by asachs01** [comment]
@jkerry this one should be g2g now:

Screen Shot 2019-07-19 at 10 41 58 AM

**Comment by quaggoth** [comment]
I don't mind taking this as-is, but I can't help but toss a recommendation everyone's way.

The time intervals use seconds. Converting from days to seconds (and backwards when decoding the use of this check) isn't ideal.

If `rootCmd.Flags().DurationVarP()` was used, it would allow the user to specify time in a more human readable way using the format supported by `time.ParseDuration()` (https://golang.org/pkg/time/#ParseDuration).

As the documentation states, this format is:
A duration string is a possibly signed sequence of decimal numbers, each with optional fraction and a unit suffix, such as "300ms", "-1.5h" or "2h45m". Valid time units are "ns", "us" (or "µs"), "ms", "s", "m", "h".

The code could take that duration and convert it to seconds for internal use.

We could follow up on this later if we want, but it would be a breaking change to the UI (like when I renamed `check_available_memory` to `check_memory`.

**Comment by asachs01** [comment]
Nice! Thanks @quaggoth for the ideas. I'm more than happy to make the change to the code to do that. It seems much easier than having to pull up a calculator every time you want to make a conversion

**Comment by quaggoth** [comment]
I'd like to see it done, but ultimately it's up to @asachs01 and @jkerry to decide this. Because I totally understand about deadlines, etc.

**Comment by asachs01** [comment]
Time format should now include the ability to use more human readable values when providing the command

**Comment by quaggoth** [comment]
Better, but use `rootCmd.Flags().DurationVarP()` for the option as in my original comment. Doing this offloads the parsing and error checking to the Cobra framework. This means invalid durations never make it to your code where you must handle it.

The code should look something like:

```
var warning time.Duration
rootCmd.Flags().DurationVarP(&warning, "warning", "w", time.Duration(72*time.Hour), "The uptime threshold to issue a warning alert, default is 72h")
```

**Comment by quaggoth** [comment]
Additionally, once you have those `time.Duration` variables populated, there's no need to use math to derive the seconds. In the above example, you would do it with:

```
warningSeconds := warning.Seconds()
```
Or better, just plop `warning.Seconds()` straight into your `GreaterFormatNagiosCheck()` call:

```
msg, retcode = nagiosformatters.GreaterFormatNagiosCheck(checkName, float64(uptime), float64(warning.Seconds()), float64(critical.Seconds()), metricName)
```

**Comment by asachs01** [comment]
AHHHHHHH! That makes much more sense now. @cwjohnston and I were getting into the weeds trying to understand how to make that happen. This clears that up! Thanks @quaggoth !

**Comment by quaggoth** [comment]
Awesome - this is looking so much better. I can clean all this up later, but if you still have time to spare...

Two more items that mostly aren't related to code but are user facing:

1. In the Check Uptime README, the example text wasn't modified after being copied from Check Memory and is still `Issue a warning if uptime usage is over 50% and critical if usage is over the default of 95%.`. Update it to match your example call.

2. Use CamelCase for the check name and set it to `CheckUptime` so it's consistent with the other checks. It's currently set to `Checkuptime`.

And now that the code is looking so much better, the code in `CheckUptime()` can be reduced to a single function call (though it's mostly a style thing). 1) Because the check name is only used once, there's no need to create a variable or const for it. 2) The extra variable declarations aren't really necessary. With this change, one can look at that function and see what it does immediately:

```
func CheckUptime(checkType string, warning, critical time.Duration, metricName string) (string, int) {
	return nagiosformatters.GreaterFormatNagiosCheck("CheckUptime", float64(getHostUptime()),
		float64(warning.Seconds()), float64(critical.Seconds()), metricName)
} 
```

Use your discretion on the code change since as I said it's mostly a style thing.

Apologies for dragging this on. We sincerely do appreciate the contribution!

**Comment by asachs01** [comment]
Sure thing! I'll take a gander at it a bit later. No worries about any sort of delay :) It's my first Golang check--I'm just happy to be making a contribution :D

**Comment by quaggoth** [comment]
Hey @asachs01, I've taken a much closer look at the `gopsutil/host` package being used to retrieve the uptime. There's another function in there, `Uptime()` that retrieves _just_ the uptime and nothing else, eliminating the overhead of populating the whole `InfoStat` structure.

We should also really be checking for errors and handling them too.  I can't imagine a situation where this would generate an error but I've also learned just about anything can happen especially when it's ran on Windows where there's such fine grained permissions.

Take a look at the following code and use it if you'd like. I've made an assumption on the requirements and am returning CRITICAL the code fails to determine the uptime.

```
// CheckUptime gathers information about the host uptime.
func CheckUptime(checkType string, warning, critical time.Duration, metricName string) (string, int) {
	const checkName = "CheckUptime"

	var msg string
	var retcode int

	uptime, err := host.Uptime()

	if err != nil {
		msg, _ = resultMessage(checkName, statusTextCritical,
			fmt.Sprintf("Failed to determine uptime: %s", err.Error()))
		retcode = 2
	} else {
		msg, retcode = nagiosformatters.GreaterFormatNagiosCheck(checkName, float64(uptime),
			float64(warning.Seconds()), float64(critical.Seconds()), metricName)
	}

	return msg, retcode
}
```

**Comment by quaggoth** [comment]
You're almost there @asachs01. Looks like after one more change I can merge it in.

In `lib/app/nagiosfoundation/check_uptime.go`, line 20 has the check name as `CheckUptime`, then line 23 has incorrect case as `CHeckUptime`. Create a `const` or a `string` variable which is set to `CheckUptime`, then use the `const` or `string` variable on those lines and it helps prevent keyboard fumbles like that.

```
const checkName = "CheckUptime"
...
msg, _ = resultMessage(checkName, statusTextCritical,
	fmt.Sprintf("Failed to determine uptime: %s", err.Error()))
...
msg, retcode = nagiosformatters.GreaterFormatNagiosCheck(checkName, float64(uptime),
	float64(warning.Seconds()), float64(critical.Seconds()), metricName)
```

File: lib/app/nagiosfoundation/check_uptime.go [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+// CheckUptime gathers information about the host uptime.
+func CheckUptime(checkType string, warning, critical time.Duration, metricName string) (string, int) {
+	
+	var msg string
+	var retcode int
+	
+	uptime, err := host.Uptime()
+	
+	if err != nil {
+		msg, _ = resultMessage("CheckUptime", statusTextCritical, fmt.Sprintf("Failed to determine uptime %s", err.Error()))
```
Comment (quaggoth) [inline lib/app/nagiosfoundation/check_uptime.go]: `CheckUptime` is correct here.

File: lib/app/nagiosfoundation/check_uptime.go [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+		retcode = 2
+	} else {
+		msg, retcode = nagiosformatters.GreaterFormatNagiosCheck("CHeckUptime", float64(uptime), float64(warning.Seconds()), float64(critical.Seconds()), metricName)
```
Comment (quaggoth) [inline lib/app/nagiosfoundation/check_uptime.go]: But `CHeckUptime` here.

File: lib/app/nagiosfoundation/check_uptime.go [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+	"fmt"
+	"time"
+		
+	"github.com/shirou/gopsutil/host"
+	"github.com/ncr-devops-platform/nagiosfoundation/lib/pkg/nagiosformatters"
+)
+
+// CheckUptime gathers information about the host uptime.
+func CheckUptime(checkType string, warning, critical time.Duration, metricName string) (string, int) {
+
```
Comment (quaggoth) [inline lib/app/nagiosfoundation/check_uptime.go]: Add a `const` or `string` for `CheckUptime` here and use them below...

**Comment by asachs01** [comment]
@quaggoth thanks again for being so patient with me on this one. I think I've got it now?

**Comment by quaggoth** [comment]
You're the one that has been so patient with me @asachs01. We both made it through and now we have a new check in the collection thanks to you.

I'll be following up with a PR containing unit tests for the API part of your check shortly. You might want to check back sometime in the future once it has bee merged to see how easy it is to add them.

**Comment by asachs01** [comment]
Awesome! I'll take a look at them for sure :)

**Comment by quaggoth** [comment]
Unit tests in PR #169

**Pull request merged**